server/service/osquery.go

```diff
@@ -1405,3 +1414,171 @@ func (svc *Service) SubmitResultLogs(ctx context.Context, logs []json.RawMessage
+}
+
+// The "snapshot" array in a ScheduledQueryResult can contain multiple rows.  Each
+// row is saved as a separate ScheduledQueryResultRow. ie. a result could contain
+// many USB Devices or a result could contain all User Accounts on a host.
+func (svc *Service) overwriteResultRows(ctx context.Context, result fleet.ScheduledQueryResult, queryID, hostID uint) error {
+	fetchTime := time.Now()
+
+	rows := make([]*fleet.ScheduledQueryResultRow, 0, len(result.Snapshot))
+
+	for _, snapshotItem := range result.Snapshot {
+
+		row := &fleet.ScheduledQueryResultRow{
+			QueryID:     queryID,
+			HostID:      hostID,
+			Data:        snapshotItem,
+			LastFetched: fetchTime,
+		}
+
+		rows = append(rows, row)
+
+	}
+
+	if err := svc.ds.OverwriteQueryResultRows(ctx, rows); err != nil {
+		return newOsqueryError("saving query result row: " + err.Error())
+	}
+
+	return nil
+}
+
+// getMostRecentResults returns only the most recent result per query.
+// Osquery can send multiple results for the same query (ie. if an agent loses
+// network connectivity it will cache multiple results).  Query Reports only
+// save the most recent result for a given query.
+func getMostRecentResults(results []fleet.ScheduledQueryResult) []fleet.ScheduledQueryResult {
+	// Use a map to track the most recent entry for each unique QueryName
+	latestResults := make(map[string]fleet.ScheduledQueryResult)
+
+	for _, result := range results {
+		if existing, ok := latestResults[result.QueryName]; ok {
+			// Compare the UnixTime time and update the map if the current result is more recent
+			if result.UnixTime > existing.UnixTime {
+				latestResults[result.QueryName] = result
+			}
+		} else {
+			latestResults[result.QueryName] = result
+		}
+	}
+
+	// Convert the map back to a slice
+	var filteredResults []fleet.ScheduledQueryResult
+	for _, v := range latestResults {
+		filteredResults = append(filteredResults, v)
+	}
+
+	return filteredResults
+}
+
+// Query names recieved from osqueryd are prefixed by teamID so we need
+// to pull them out to match the query name and team ID in the database
+func getQueryNameAndTeamIDFromResult(path string) (*uint, string, error) {
+	// For pattern: pack/Global/Name
+	if strings.HasPrefix(path, "pack/Global/") {
+		return nil, lastItemFromPath(path), nil
+	}
+
+	// For pattern: pack/team-<ID>/Name
+	if strings.HasPrefix(path, "pack/team-") {
+		parts := strings.SplitN(path, "/", 3)
+		if len(parts) != 3 {
+			return nil, "", fmt.Errorf("unknown format: %s", path)
+		}
+
+		teamNumberStr := strings.TrimPrefix(parts[1], "team-")
+		teamNumberUint, err := strconv.ParseUint(teamNumberStr, 10, 32)
+		if err != nil {
+			return nil, "", fmt.Errorf("parsing team number: %w", err)
+		}
+
+		teamNumber := uint(teamNumberUint)
+		return &teamNumber, parts[2], nil
+	}
+
+	// For pattern: pack/PackName/Query (legacy pack)
+	if strings.Count(path, "/") == 2 && strings.HasPrefix(path, "pack/") {
```

Comment:
I accidentally added the comment on the commit.

This won't catch queries with `/` in their name. (Seems we support such a thing.)